### DISCUSSION TITLE
Bulk export relation record

### DISCUSSION THREAD

**Discussion by huiyang** [filamentphp/filament#13549] [Help]
### Package

Actions

### Package Version

v3.2.23

### How can we help you?

I have a built a list/datatable for cakes, and i want to add bulk action to allow user to export cake list, as well as cakes steps.
for cakes list, it is no issue and very straight forward. But when i want to export cakes steps, which is related records of cakes, i not sure how can i do?

I tried to use the records() method to let the filament ExportBulkAction to get cake steps as records instead of cakes, it just not working and ignored (overwrite) whatever i to set. (as explained below)

```
// Bulk action to export cakes list (no issue)
ExportBulkAction::make()->color('primary')
                    ->exporter(CakeExporter::class), 
// Bulk action to export cakes steps 
                \App\Filament\Actions\ExportBulkAction::make('export_steps')->color('primary')
                    ->label('Export cake steps')
                    ->exporter(CakeStepExporter::class)
                    ->records(function() {
                         // don's have any effect for whatever i try to do here
                    })
                    ,
```

So what is the best way i can do? Do i really need to extends the ExportBulkAction class to archive my purpose?

Thanks

**Comment by bomshteyn** [comment]
AFIK there is no `records` method available on `ExportBulkAction`. 

I have had a similar need and got it mostly working. In my scenario the client wanted to have an admin order export from the order table view. But since there could be multiple order items in each order the client wanted the export file to have a new row for each order item / product. So pretty similar to your scenario.

I have managed to make it happen except for one problem that the columns are now not in the right order right now, not sure why this is causing this issue.


I have added `modifyQueryUsing` line to the export action

```PHP
ExportAction::make()
                ->label('Export Orders')
                ->icon('heroicon-o-arrow-down-tray')
                ->modifyQueryUsing(function ($query) {
                    return OrderItem::query()
                       ->whereIn('order_id', $query->get('id')->pluck('id'));
                })
                ->exporter(AdminOrderExporter::class),
```

And inside my `AdminOrderExporter` class, I have changed the model

```PHP
// protected static ?string $model = Order::class;
protected static ?string $model = OrderItem::class;
```

It would be really cool if you can make your own iterators inside `public static function getColumns(): array` and return an array of arrays for multi-row results

**Comment by GarethSomers** [comment]
Hi @bomshteyn does this still work ? Couldn't get your solution to work (v3.3)

**Reply by aizensoosuke** [reply]
Unfortunately @bomshteyn uses non-bulk actions for that workaround.  
  
The workaround does not work with bulk actions, because `instanceof` is used to straight up ignore the query when the action used is `ExportBulkAction`.  
  
One workaround would be to use `tap` to modify `$this->records`, but it's a protected property so it's not possible either. And we cannot extend the `ExportBulkAction` model, because the type `ExportAction | ExportTableAction | ExportTableBulkAction` is enforced in the file `CanExportRecords.php`.  
  
So, short of copy-pasting the whole thing, here's a workaround that's quite less user friendly but does the job. In my case, I want a button "Export Users" on my "Groups" table, that exports the list of selected groups' members.

First, create a filter on the users table that filters by multiple groups.
```php
// app/Filament/Resources/UserResource.php
Tables\Filters\SelectFilter::make('Groups')  
    ->multiple()  
    ->relationship('groups', 'name'),
```

Then, add a custom bulk action to the groups table that redirects to the pre-filtered user table.
```php
// app/Filament/Resources/GroupResource.php
Tables\Actions\BulkAction::make('export_members')
    ->label('Export members')
    ->color('gray')
    ->icon('heroicon-o-arrow-down-tray')
    ->modalDescription('You will be taken to the list of all the selected groups\' users. Use the button "Export Users" at the top right of the next page.')
    ->modalSubmitActionLabel('Go')
    ->action(function (Collection $records) {
        $params = $records
            ->mapWithKeys(function (Group $group, int $i) {
                return ["tableFilters[Groups][values][{$i}]" => $group->id];
            })
            ->all();
    
        return redirect(UserResource::getUrl('index', $params));
    }),
```

Make sure you have the header export action in your users resource:
```php
// app/Filament/Resources/UserResource.php
->headerActions([  
    ExportAction::make()   
        ->exporter(UserExporter::class),  
])
```

**Comment by GarethSomers** [comment]
Hey @aizensoosuke 

I appreciate your input here.

I revisited this today and found a solution, although a little messy works within the pre-existing UX (no filters required).

First of all I extended `\Filament\Tables\Actions\ExportBulkAction` and created my own `App\Filament\Tables\Actions\OrderItemExportBulkAction`.

<details>

<summary>Show / Hide OrderItemExportBulkAction</summary>

```
<?php

namespace App\Filament\Tables\Actions;

use Filament\Actions\Concerns\CanExportRecords;
use Illuminate\Database\Eloquent\Collection as EloquentCollection;
use Illuminate\Support\Collection;

class OrderItemExportBulkAction extends \Filament\Tables\Actions\ExportBulkAction
{
    public function getRecords(): EloquentCollection | Collection | null
    {
        $results = \App\Models\OrderItem::whereIn('order_id', $this->evaluate($this->records)->toArray())->get()->pluck('id');

        return $results;
    }

    public static function modifyQuery(Builder $query): Builder
    {
        $ids = $query->toArray();

        $results = \App\Models\Models\OrderItem::whereIn('order_id', $ids)->get()->toArray();

        return $results;
    }
}
```

</details>

Ensuring to use it within my resource:
<details>

<summary>Show / Hide Resource</summary>

```
\App\Filament\Tables\Actions\ExportBulkAction::make('export_order_items')
    ->label('Export Orders Item(s)')
    ->exporter(\App\Filament\Exports\OrderItemExporter::class)
```

</details>

Then within my OrderItemExporter I've added the `modifyQuery` method:

<details>

<summary>Show / Hide OrderItemExporter</summary>

```
<?php

namespace App\Filament\Exports;

use Carbon\CarbonInterface;
use Illuminate\Database\Eloquent\Model;
use App\Models\Order;
use App\Models\OrderItem;
use Filament\Actions\Exports\ExportColumn;
use Filament\Actions\Exports\Exporter;
use Filament\Actions\Exports\Models\Export;
use Illuminate\Database\Eloquent\Builder;

class OrderItemExporter extends Exporter
{
    protected static ?string $model = OrderItem::class;

    public static function getColumns(): array
    {
        return [
            ExportColumn::make('order.id')
                ->label('Order ID'),
            ExportColumn::make('id')
                ->label('Order Item ID')
        ];
    }

    public static function modifyQuery(Builder $query): Builder
    {
        return OrderItem::whereIn('order_id', $query->get()->pluck('id')->toArray());
    }
}
```

</details>
